File: src/openrgb_flowers/math/easing.py
"""Smooth easing and harmonic mathematical functions."""
from __future__ import annotations
import numpy as np


class EasingFunctions:
    """Mathematical smoothing and harmonic functions for organic natural animations."""

    @staticmethod
    def smoothstep(edge0: float, edge1: float, x: np.ndarray) -> np.ndarray:
        """GLSL-equivalent smoothstep interpolation for smooth edge falloffs."""
        diff = edge1 - edge0
        if abs(diff) < 1e-9:
            return np.where(x >= edge1, 1.0, 0.0).astype(np.float32)
        t = np.clip((x - edge0) / diff, 0.0, 1.0)
        return t * t * (3.0 - 2.0 * t)

    @staticmethod
    def ease_out_cubic(t: float | np.ndarray) -> float | np.ndarray:
        """Cubic deceleration curve."""
        return 1.0 - np.power(1.0 - t, 3.0)

    @staticmethod
    def ease_in_out_cubic(t: float | np.ndarray) -> float | np.ndarray:
        """Smooth S-curve acceleration and deceleration."""
        return np.where(t < 0.5, 4.0 * t * t * t, 1.0 - np.power(-2.0 * t + 2.0, 3.0) / 2.0)

    @staticmethod
    def petal_harmonic(
        angles: np.ndarray,
        lobes: int,
        depth: float,
        rotation: float,
    ) -> np.ndarray:
        """Calculates multi-petal polar modulation factor: (1.0 + depth * cos(lobes * theta + rot))."""
        return 1.0 + depth * np.cos(lobes * angles + rotation)

File: src/math/easing.cpp
#include "openrgb_flowers/math/easing.hpp"

#include <algorithm>
#include <cmath>

namespace openrgb_flowers::math {

float EasingFunctions::smoothstep(float edge0, float edge1, float x) {
    float diff = edge1 - edge0;
    if (std::fabs(diff) < 1e-9f) {
        return (x >= edge1) ? 1.0f : 0.0f;
    }
    float t = std::clamp((x - edge0) / diff, 0.0f, 1.0f);
    return t * t * (3.0f - 2.0f * t);
}

float EasingFunctions::ease_out_cubic(float t) {
    float inv = 1.0f - std::clamp(t, 0.0f, 1.0f);
    return 1.0f - inv * inv * inv;
}

float EasingFunctions::ease_in_out_cubic(float t) {
    float clamped = std::clamp(t, 0.0f, 1.0f);
    if (clamped < 0.5f) {
        return 4.0f * clamped * clamped * clamped;
    }
    float f = -2.0f * clamped + 2.0f;
    return 1.0f - (f * f * f) / 2.0f;
}

float EasingFunctions::petal_harmonic(float angle, int lobes, float depth, float rotation) {
    return 1.0f + depth * std::cos(static_cast<float>(lobes) * angle + rotation);
}

} // namespace openrgb_flowers::math
